### PR TITLE
Feature/keyboard

### PR DESCRIPTION
**Pull request by Tedy2606**
finished keyboard driver and interrupts

### CODE REVIEW

File: cpu/timer/timer.c [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+#define MODE_COMMAND_REGISTER 0x43
+#define CHANNEL0_PORT 0x40
+
+uint32 ticks = 0;
+
+/**********************************************************
+ * Private kernel functions signiture                     *
+ **********************************************************/
+
+static void timer_handler(interrupt_registers_struct regs);
```
Comment (ShonGutman) [inline cpu/timer/timer.c:19]: no need for static

File: libc/string.c [hunk 2]
```diff
@@ -0,0 +1,79 @@
+#include "string.h"
+
+
+/**
+ * K&R implementation
+ */
+void int_to_ascii(int n, char *str)
```
Comment (ShonGutman) [inline libc/string.c:7]: const? uint32? use them

File: libc/string.c [hunk 3]
```diff
@@ -0,0 +1,79 @@
+#include "string.h"
+
+
+/**
+ * K&R implementation
+ */
+void int_to_ascii(int n, char *str) 
+{
+    int i, sign;
```
Comment (ShonGutman) [inline libc/string.c:9]: we init vars like this `uint32 i = 0`

File: libc/string.c [hunk 4]
```diff
@@ -0,0 +1,79 @@
+#include "string.h"
+
+
+/**
```
Comment (ShonGutman) [inline libc/string.c:4]: delete lines 4-6

File: libc/string.c [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+void backspace(char *str) 
+{
+    int len = strlen(str);
+    str[len-1] = STRING_TERMINATOR;
+}
+
+/* K&R 
+ * Returns <0 if s1<s2, 0 if s1==s2, >0 if s1>s2 */
+int strcmp(const char *str1, const char *str2)
```
Comment (ShonGutman) [inline libc/string.c:68]: pointer in right side :( it ok but i do not like it

File: libc/string.c [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    int len = strlen(str);
+    str[len-1] = STRING_TERMINATOR;
+}
+
+/* K&R 
+ * Returns <0 if s1<s2, 0 if s1==s2, >0 if s1>s2 */
+int strcmp(const char *str1, const char *str2) 
+{
+    int i;
+    for (i = 0; str1[i] == str2[i]; i++)
```
Comment (ShonGutman) [inline libc/string.c:71]: again uint32 pls and do it in for loop...

File: libc/string.h [hunk 7]
```diff
@@ -0,0 +1,51 @@
+#ifndef STRINGS_H
+#define STRINGS_H
+
+#define STRING_TERMINATOR '\0'
```
Comment (ShonGutman) [inline libc/string.h:4]: definei in strings.c pls, cause when we include this header file it will be included for no reason

File: libc/string.h [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+/*
+ *  turns an int into ascii
+ * @param n - the number to turn to ascii
+ * @param str - the str the number is going to be in
+ * @return The str with the number in it 
+ */
+void int_to_ascii(int n, char *str);
```
Comment (ShonGutman) [inline libc/string.h:12]: change name to numer

File: libc/string.h [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+/*
+ *  reverses a string
+ * @param str - the str to be reveresd
+ * @return The reversed str
```
Comment (ShonGutman) [inline libc/string.h:17]: no return value

File: libc/string.h [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+/*
+ *  reverses a string
+ * @param str - the str to be reveresd
+ * @return The reversed str  
+ */
+void reverse(char *str);
+
+/*
+ *  checks string lenght
```
Comment (ShonGutman) [inline libc/string.h:22]: function doc is not good. function returns string *length*

**Review by ShonGutman — COMMENTED** [review]
for now

File: libc/string.h (hunk 7 above)
Comment (Tedy2606) [reply libc/string.h:4]: this define is also used in other files so it should stay in the header I belive

File: libc/string.h (hunk 8 above)
Comment (Tedy2606) [reply libc/string.h:12]: number*

File: drivers/keyboard.c [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+ * 
+ *  -------------------------------------
+ */
+
+//maximum amount of chars that can be inputed in one line.
+
+static char key_buffer[MAX_KEYS];
+
+int capslock_value_change = CAPPED_TO_NORMAL_ASCII;
+int shift_pressed = 0;
```
Comment (ShonGutman) [inline drivers/keyboard.c:68]: it is important to follow and use our types. in this case - uint8
moreover it is important to mark them as static, I will explain to you why if you want

File: drivers/keyboard.c [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+void keyboard_handler(interrupt_registers_struct regs) 
+{
+    /* The PIC leaves us the scancode in port 0x60 */
+    uint8 scancode = port_byte_in(0x60);
```
Comment (ShonGutman) [inline drivers/keyboard.c:73]: please make this in define like this `#define KEYBOARD_DATA_PORT 0x60`

File: drivers/keyboard.c [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+    //if scancode is unknow then return
```
Comment (ShonGutman) [inline drivers/keyboard.c:74]: type - unknown

File: drivers/keyboard.c [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+            letter = scancode_ascii[(int)scancode];
+        }
+        // we dont want to change anything other than letters
+        if (letter != SPACE_ASCII && !shift_pressed && letter >= ASCII_ABC_START && letter <= ASCII_ABC_END)
+        {
+            letter += capslock_value_change;
+        }
+        char str[2] = {letter, STRING_TERMINATOR};
+        append(key_buffer, letter);
+        print(str);
```
Comment (ShonGutman) [inline drivers/keyboard.c:145]: make print_char public and use this func

**Review by ShonGutman — COMMENTED** [review]
changes to be made

File: drivers/keyboard.c [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+    {
+        char letter = 0;
+        if (shift_pressed)
+        {
+            letter = scancode_ascii_shifted[(int)scancode];
+        }
+        else
+        {
+            letter = scancode_ascii[(int)scancode];
+        }
```
Comment (ShonGutman) [inline drivers/keyboard.c:137]: again, use uint32 please

**Review by ShonGutman — APPROVED** [review]
Well done

**Pull request merged**